### PR TITLE
Add StreamManager

### PR DESCRIPTION
**Pull request by verult**
The full StreamManager implementation prototyped in https://github.com/quantumlib/Cirq/pull/6145.

Some tasks are deferred:
- Add the ability to close the request iterator in order to close the stream. Being able to stop as soon as receiving the signal rather than waiting for the next request is non-trivial. Since it's not core to streaming functionality, this will be a follow-up soon after MVP.
- Other TODOs are less critical.

This PR depends on the latest version of duet which has not been released. It contains fixes for `AwaitableFuture` cancellations.

I'm planning to add a README with design and sequence diagrams. Please let me know if you'd like to hold off on reviewing this PR until that's ready.

Only the last commit needs to be reviewed. All other commits are from https://github.com/quantumlib/Cirq/pull/6190.

@maffoo @wcourtney

### CODE REVIEW

**Comment by verult** [comment]
Btw, if there's a way to set a PR as diffbase, please let me know!

**Comment by verult** [comment]
My recommendation is to review methods with long docstrings first - they are the most important pieces.

**Comment by verult** [comment]
Fixed a race involving `StreamManager.stop()`: it cancels the stream coroutine, but cancellation is handled in the asyncio thread, and the duet thread does not wait for the cancellation to finish. I fixed it by handling cancellation as a duet future callback.

The lingering cancellation failure is because the latest fixes in duet is not yet released.

File: cirq-google/cirq_google/engine/stream_manager.py [hunk 1]
```diff
@@ -59,5 +80,222 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
...
+
+        current_request = create_program_and_job_request
+        while True:
+            try:
+                current_request.message_id = self._generate_message_id()
+                response_future = self._response_demux.subscribe(current_request.message_id)
+                await self._request_queue.put(current_request)
+                response = await response_future
+
+            # Broken stream
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/stream_manager.py:227]: Have you verified that broken streams look like this?  e.g. by terminating a locally running server or messing up your connection to a prod instance (or better yet - with an e2e test)?

File: cirq-google/cirq_google/engine/stream_manager.py [hunk 2]
```diff
@@ -59,5 +80,222 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
...
+                    # Retry
+                    self._response_demux.unsubscribe(current_request.message_id)
+                    current_request = get_result_request
+                    continue
+                    # TODO(#5996) add exponential backoff
+                raise e
+
+            # Either when this request is canceled or the _manage_stream() loop is canceled.
+            except asyncio.CancelledError:
+                if response_future is not None:
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/stream_manager.py:241]: ...and not completed? Else you'll get another exception here if that case is tripped.

File: cirq-google/cirq_google/engine/stream_manager.py [hunk 3]
```diff
@@ -59,5 +80,222 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
...
+                    await self._cancel(job.name)
+                raise
+
+            # Response handling
+            if 'result' in response:
+                return response.result
+            elif 'job' in response:
+                return response.job
+            elif 'error' in response:
+                current_request = _decide_retry_request_or_raise(
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/stream_manager.py:252]: Nit: `s/decide/get`

File: cirq-google/cirq_google/engine/stream_manager.py [hunk 4]
```diff
@@ -59,5 +80,222 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
...
+            except google_exceptions.GoogleAPICallError as e:
+                if _is_retryable_error(e):
+                    # Retry
+                    self._response_demux.unsubscribe(current_request.message_id)
+                    current_request = get_result_request
+                    continue
+                    # TODO(#5996) add exponential backoff
+                raise e
+
+            # Either when this request is canceled or the _manage_stream() loop is canceled.
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/stream_manager.py:237]: > either when this request is canceled or...

What's the client interface to cancel a request?

File: cirq-google/cirq_google/engine/stream_manager.py [hunk 5]
```diff
@@ -59,5 +80,222 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
...
+    error: quantum.StreamError,
+    current_request,
+    create_program_and_job_request,
+    create_job_request: quantum.QuantumRunStreamRequest,
+):
+    """Decide whether the given stream error is retryable.
+
+    If it is, returns the next stream request to send upon retry. Otherwise, raises an error.
+    """
+    if error.code == Code.PROGRAM_DOES_NOT_EXIST:
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/stream_manager.py:284]: Do we have Python 3.10 for match/case statements? https://docs.python.org/3/whatsnew/3.10.html#pep-634-structural-pattern-matching

File: cirq-google/cirq_google/engine/stream_manager.py [hunk 6]
```diff
@@ -59,5 +80,222 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
...
+        """
+
+        # Construct requests ahead of time to be reused for retries.
+        create_program_and_job_request = quantum.QuantumRunStreamRequest(
+            parent=project_name,
+            create_quantum_program_and_job=quantum.CreateQuantumProgramAndJobRequest(
+                parent=project_name, quantum_program=program, quantum_job=job
+            ),
+        )
+        create_job_request = quantum.QuantumRunStreamRequest(
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/stream_manager.py:205]: Does this get used as a retry request type somewhere? I didn't find it ctrl+f'ing around.

File: cirq-google/cirq_google/engine/stream_manager.py [hunk 7]
```diff
@@ -59,5 +80,222 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
...
+    if error.code == Code.PROGRAM_DOES_NOT_EXIST:
+        if 'create_quantum_job' in current_request:
+            return create_program_and_job_request
+    elif error.code == Code.PROGRAM_ALREADY_EXISTS:
+        if 'create_quantum_program_and_job' in current_request:
+            raise ProgramAlreadyExistsError(
+                current_request.create_quantum_program_and_job.quantum_program.name
+            )
+    elif error.code == Code.JOB_DOES_NOT_EXIST:
+        if 'get_quantum_result' in current_request:
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/stream_manager.py:293]: Can you talk through this error mechanism? The best cause I can imagine is that the job was deleted, but then we shouldn't create a new one.

If that's true, then we wouldn't call `create_job_request` anywhere, which also makes the`PROGRAM_DOES_NOT_EXIST` error handling obsolete.

**Review by wcourtney — CHANGES_REQUESTED** [review]
First pass % tests. Really great overall, mostly nits and edge cases to consider. Thanks for all the attention to detail!

File: cirq-google/cirq_google/engine/stream_manager.py [hunk 8]
```diff
@@ -59,5 +80,222 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
...
+    the stream and returns a future to be completed when either the result is ready or the job has
+    failed.
+
+    A new stream is opened during the first `submit()` call, and it stays open. If the stream is
+    unused, users could close the stream and free management resources by calling `stop()`.
+    """
+
+    def __init__(self, grpc_client: quantum.QuantumEngineServiceAsyncClient):
+        self._grpc_client = grpc_client
+        self._request_queue: asyncio.Queue = asyncio.Queue()
```
[the remark this answers is not in the document]
Comment (verult) [reply cirq-google/cirq_google/engine/stream_manager.py]: I used that approach at first, but I got an error somewhere. I think it was a mypy error about `self._response_demux` being uninitialized before being read. Seems like assignments in `__init__()` is regarded as special here.

File: cirq-google/cirq_google/engine/stream_manager.py [hunk 9]
```diff
@@ -59,5 +80,222 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
...
+            try:
+                # The default gRPC client timeout is used.
+                response_iterable = await self._grpc_client.quantum_run_stream(
+                    self._request_iterator()
+                )
+                async for response in response_iterable:
+                    self._response_demux.publish(response)
+            except BaseException as e:
+                # TODO(#5996) Close the request iterator to close the existing stream.
+                # Note: the message ID counter is not reset upon a new stream.
```
[the remark this answers is not in the document]
Comment (verult) [reply cirq-google/cirq_google/engine/stream_manager.py:175]: Thanks for the catch. We actually probably shouldn't reset the ID to avoid potential for any collisions. For example, the response to a request from a previous stream may be sent over the new stream.

File: cirq-google/cirq_google/engine/stream_manager.py [hunk 10]
```diff
@@ -59,5 +80,222 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
...
+            A future for the job result, or the job if the job has failed.
+
+        Raises:
+            ProgramAlreadyExistsError if the program already exists.
+            StreamError if there is a non-retryable error while executing the job.
+            concurrent.futures.CancelledError if the stream is stopped while a job is in flight.
+            google.api_core.exceptions.GoogleAPICallError if the stream breaks with a non-retryable
+                error.
+        """
+        if self._manage_stream_loop_future is None:
```
[the remark this answers is not in the document]
Comment (verult) [reply cirq-google/cirq_google/engine/stream_manager.py:125]: We should also leave it alone. Thanks for the catch - updated.

File: cirq-google/cirq_google/engine/stream_manager.py [hunk 11]
```diff
@@ -59,5 +80,222 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
...
+
+    def stop(self) -> None:
+        """Closes the open stream and resets all management resources."""
+        if self._manage_stream_loop_future is not None:
+            self._manage_stream_loop_future.cancel()
+        self._reset()
+
+    def _reset(self):
+        """Resets the manager state."""
+        self._request_queue: asyncio.Queue = asyncio.Queue()
```
[the remark this answers is not in the document]
Comment (verult) [reply cirq-google/cirq_google/engine/stream_manager.py]: Cirq still supports Python 3.9, where `Queue` isn't generic.

File: cirq-google/cirq_google/engine/stream_manager.py [hunk 12]
```diff
@@ -59,5 +80,222 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
...
+                    self._request_iterator()
+                )
+                async for response in response_iterable:
+                    self._response_demux.publish(response)
+            except BaseException as e:
+                # TODO(#5996) Close the request iterator to close the existing stream.
+                # Note: the message ID counter is not reset upon a new stream.
+                self._response_demux.publish_exception(e)  # Raise to all request tasks
+                if isinstance(e, asyncio.CancelledError):
+                    break
```
[the remark this answers is not in the document]
Comment (verult) [reply cirq-google/cirq_google/engine/stream_manager.py]: I don't think so - the caller triggering the cancellation (`StreamManager.stop()`) already resets, so it would be redundant.

File: cirq-google/cirq_google/engine/stream_manager.py [hunk 13]
```diff
@@ -59,5 +80,222 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
...
+        This coroutine sends QuantumRunStream requests to the request iterator and receives
+        responses from the ResponseDemux.
+
+        It initially sends a CreateQuantumProgramAndJobRequest, and retries if there is a retryable
+        error by sending another request. The exact request type depends on the error.
+
+        There is one execution coroutine per running job submission.
+        """
+
+        # Construct requests ahead of time to be reused for retries.
```
[the remark this answers is not in the document]
Comment (verult) [reply cirq-google/cirq_google/engine/stream_manager.py:198]: Thanks for the suggestion. Updated

File: cirq-google/cirq_google/engine/stream_manager.py [hunk 14]
```diff
@@ -59,5 +80,222 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
...
+            create_quantum_job=quantum.CreateQuantumJobRequest(
+                parent=program.name, quantum_job=job
+            ),
+        )
+        get_result_request = quantum.QuantumRunStreamRequest(
+            parent=project_name, get_quantum_result=quantum.GetQuantumResultRequest(parent=job.name)
+        )
+
+        current_request = create_program_and_job_request
+        while True:
```
[the remark this answers is not in the document]
Comment (verult) [reply cirq-google/cirq_google/engine/stream_manager.py:220]: I can do that in a follow up while implementing exponential backoff. WDYT?

File: cirq-google/cirq_google/engine/stream_manager.py (hunk 4 above)
Comment (verult) [reply cirq-google/cirq_google/engine/stream_manager.py:236]: Clarified in the class and `submit()` docstrings.

File: cirq-google/cirq_google/engine/stream_manager.py (hunk 5 above)
Comment (verult) [reply cirq-google/cirq_google/engine/stream_manager.py:283]: Good to know this feature is in Python now! Unfortunately Cirq still supports 3.9.

File: cirq-google/cirq_google/engine/stream_manager.py (hunk 6 above)
Comment (verult) [reply cirq-google/cirq_google/engine/stream_manager.py:205]: It's used in `_decide_retry_request_or_raise()` when there is a `StreamError`.

File: cirq-google/cirq_google/engine/stream_manager.py (hunk 7 above)
Comment (verult) [reply cirq-google/cirq_google/engine/stream_manager.py:292]: Good point. What about the case where `CreateProgramAndJobRequest` isn't sent properly before the stream breaks? A `GetResultRequest` will be issued but the job doesn't exist. Unfortunately this is indistinguishable from the case where the job was deleted. I would say it's OK for the job to be recreated because if a user truly wants to stop a job execution, they could cancel the future returned by `submit()`. WDYT?

File: cirq-google/cirq_google/engine/stream_manager.py [hunk 15]
```diff
@@ -59,5 +80,222 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
...
+        create_program_and_job_request = quantum.QuantumRunStreamRequest(
+            parent=project_name,
+            create_quantum_program_and_job=quantum.CreateQuantumProgramAndJobRequest(
+                parent=project_name, quantum_program=program, quantum_job=job
+            ),
+        )
+        create_job_request = quantum.QuantumRunStreamRequest(
+            parent=project_name,
+            create_quantum_job=quantum.CreateQuantumJobRequest(
+                parent=program.name, quantum_job=job
```
[the remark this answers is not in the document]
Comment (verult) [reply cirq-google/cirq_google/engine/stream_manager.py:208]: Hmm, the only reason why `program.name` is currently required by `StreamManager` is that `StreamError` doesn't return the program/job that already exists or doesn't exist. That information is likely in the error message, but we can't rely on that.

So I think the caller has to randomly generate a program name. Updated.

**Comment by maffoo** [comment]
@verult, can you rebase on or merge with master to resolve the conflicts? `ResponseDemux is still showing up in the diff but that was already merged in #6190.

File: cirq-google/cirq_google/engine/stream_manager.py (hunk 1 above)
Comment (verult) [reply cirq-google/cirq_google/engine/stream_manager.py:227]: I have not, thanks for the great catch. Will do an experiment in parallel with this PR.

File: cirq-google/cirq_google/engine/stream_manager.py (hunk 2 above)
Comment (verult) [reply cirq-google/cirq_google/engine/stream_manager.py:240]: Cancelling an already cancelled future returns False instead of raising. It would be good practice to check the future is not done, but it actually requires a bigger change that I don't think I should do now: if `response_future` is cancelled, it is also `done()`, and to fix this, it would be best to move the future cancellation logic into a callback instead of processing it here as an exception, but we need extra changes to somehow pass the job name into the callback.

**Comment by verult** [comment]
Addressed Will's comments and rebased with master.

File: cirq-google/cirq_google/engine/stream_manager.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (wcourtney) [reply cirq-google/cirq_google/engine/stream_manager.py:152]: It doesn't look like this is updated?

File: cirq-google/cirq_google/engine/stream_manager.py [hunk 16]
```diff
@@ -134,27 +141,35 @@ def submit(
...
+        try:
+            if (
+                self._manage_stream_loop_future is not None
+                and not self._manage_stream_loop_future.done()
+            ):
+                self._manage_stream_loop_future.cancel()
+        finally:
+            self._reset()
 
     def _reset(self):
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/stream_manager.py:168]: Should "reset" cancel actively running tasks (call `stop()`)? That would cause a loop right now, but it makes sense that we'd clear old work as part of a reset.

File: cirq-google/cirq_google/engine/stream_manager.py [hunk 17]
```diff
@@ -134,27 +141,35 @@ def submit(
...
-            self._manage_stream_loop_future.cancel()
-        self._reset()
+        try:
+            if (
+                self._manage_stream_loop_future is not None
+                and not self._manage_stream_loop_future.done()
+            ):
+                self._manage_stream_loop_future.cancel()
+        finally:
+            self._reset()
```
Comment (wcourtney) [inline cirq-google/cirq_google/engine/stream_manager.py:166]: What if you just want to stop the `StreamManager` altogether because you're done with it? WDYT of separating `reset` from `stop` or at least making it optional?

**Review by wcourtney — APPROVED** [review]
LGTM % some smallish suggestions. Thanks!!! 🙌

File: cirq-google/cirq_google/engine/stream_manager.py (hunk 16 above)
Comment (verult) [reply cirq-google/cirq_google/engine/stream_manager.py:168]: Chatted offline - this no longer applies

File: cirq-google/cirq_google/engine/stream_manager.py (hunk 17 above)
Comment (verult) [reply cirq-google/cirq_google/engine/stream_manager.py:166]: Chatted offline - this no longer applies

File: cirq-google/cirq_google/engine/stream_manager.py [hunk 18]
```diff
@@ -75,5 +96,276 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
...
+                response = await response_future
+
+            # Broken stream
+            except google_exceptions.GoogleAPICallError as e:
+                if _is_retryable_error(e):
+                    # Retry
+                    current_request = _to_get_result_request(create_program_and_job_request)
+                    continue
+                    # TODO(#5996) add exponential backoff
+                raise e
```
Comment (maffoo) [inline cirq-google/cirq_google/engine/stream_manager.py:257]: nit: I'd prefer to handle the error case first, e.g.
```suggestion
                if not _is_retryable_error(e):
                    raise e
                    
                # Retry
                current_request = _to_get_result_request(create_program_and_job_request)
                continue
                # TODO(#5996) add exponential backoff
```
Could also consider putting the success case below in an `else` clause of the `try-except`, which would avoid the need for the `continue`.

**Review by maffoo — COMMENTED** [review]
Some minor comments, but mostly looks good. I still have some concerns about both the duet and asyncio threads mutating the StreamManager without synchronization, but I think we can address those issues in future PRs.

**Review by maffoo — APPROVED** [review]
